python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,502 @@
+
+    @to.register(DecimalType)
+    def _(self, type_var: DecimalType) -> "DecimalLiteral":
+        if type_var.scale == 0:
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal("1."), rounding=ROUND_HALF_UP))
+        else:
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal((0, (1,), -type_var.scale)), rounding=ROUND_HALF_UP))
+
+
+class DoubleLiteral(Literal[float]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(DoubleType)
+    def _(self, type_var: DoubleType) -> "DoubleLiteral":
+        return self
+
+    @to.register(FloatType)
+    def _(self, type_var: FloatType) -> Union[AboveMax, BelowMin, FloatLiteral]:
+        if FloatType.max < self.value:
+            return AboveMax()
+        elif FloatType.min > self.value:
+            return BelowMin()
+        return FloatLiteral(self.value)
+
+    @to.register(DecimalType)
+    def _(self, type_var: DecimalType) -> "DecimalLiteral":
+        if type_var.scale == 0:
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal("1."), rounding=ROUND_HALF_UP))
+        else:
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal((0, (1,), -type_var.scale)), rounding=ROUND_HALF_UP))
+
+
+class DateLiteral(Literal[int]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(DateType)
+    def _(self, type_var: DateType) -> "DateLiteral":
+        return self
+
+
+class TimeLiteral(Literal[int]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(TimeType)
+    def _(self, type_var: TimeType) -> "TimeLiteral":
+        return self
+
+
+class TimestampLiteral(Literal[int]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(TimestampType)
+    def _(self, type_var: TimestampType) -> "TimestampLiteral":
+        return self
+
+    @to.register(DateType)
+    def _(self, type_var: DateType) -> "DateLiteral":
+        return DateLiteral((datetime.datetime.fromtimestamp(self.value / 1000000) - EPOCH).days)
+
+
+class DecimalLiteral(Literal[Decimal]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(DecimalType)
+    def _(self, type_var: DecimalType) -> Optional["DecimalLiteral"]:
+        if type_var.scale == abs(self.value.as_tuple().exponent):
+            return self
+        return None
+
+
+class StringLiteral(Literal[str]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(StringType)
+    def _(self, type_var: StringType) -> "StringLiteral":
+        return self
+
+    @to.register(DateType)
+    def _(self, type_var: DateType) -> "DateLiteral":
+        return DateLiteral((datetime.datetime.strptime(self.value, "%Y-%m-%d") - EPOCH).days)
+
+    @to.register(TimeType)
+    def _(self, type_var: TimeType) -> "TimeLiteral":
```

Comment:
Similar to the other date/time types, this can be done a bit more easily:

```python
t = time.fromisoformat(self.value)
return (((t.hour * 60 + t.minute) * 60) + t.second) * 1_000_000 + t.microsecond
```

The `time.fromisoformat` method does a good job of validating the format, so no additional regex is needed.

Reply:
Thank you Ryan, your example in String to TimeStamp/Date/Time is really helpful